creator/ingest_runs/models.py

```diff
@@ -73,6 +82,39 @@ class Meta:
         help_text="The user who submitted this ingest run",
     )
 
+    state = FSMField(
+        default="waiting", help_text="The current state of the ingest run"
+    )
+
+    @transition(field=state, source="waiting", target="started")
+    def start(self):
+        """
+        Begin running the ingest process.
+        """
+        return
+
+    @transition(field=state, source="started", target="completed")
+    def complete(self):
+        """
+        Finish running the ingest process without error.
+        """
+        return
+
+    @transition(field=state, source="started", target="canceled")
+    def cancel(self):
+        """
+        Cancel the ingest run on request from the user or on account of a new
+        ingest run being started for the same input parameters.
+        """
+        stop_job(str(self.id), queue=IngestRun.get_job_queue(), delete=True)
+
+    @transition(field=state, source=["started", "canceled"], target="failed")
+    def fail(self):
```

Comment:
The only time a run will be in waiting state is when you create it which happens here: https://github.com/kids-first/kf-api-study-creator/blob/1f5b459ea1d3896bb42da928d36856ea4732db41/creator/ingest_runs/mutations.py#L57. Then it will immediately change to started state when the start ingest task executes. With the way things work right now I don't think there is a way for the frontend to cancel the run when its in waiting state. 

But I guess it doesn't hurt to add "waiting" to this list here and that would set us up for a future where a user could queue up a ingest run but then decide to cancel it before they start it.

Reply:
What if the queue is down or redis gets restarted or the job gets lost or the workers aren't consuming? I would think the user may want to cancel if it's been stuck in the waiting state for too long or perhaps some automated cleanup system would scan for things that have been in waiting for too long and cancel the long running ones.